src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/TupleTypeDecoder.vb

```diff
@@ -0,0 +1,308 @@
+            Return definition.Construct(subst)
+        End Function
+
+        Private Function DecodeTypeArguments(typeArgs As ImmutableArray(Of TypeSymbol)) As ImmutableArray(Of TypeSymbol)
+            If typeArgs.IsEmpty Then
+                Return typeArgs
+            End If
+
+            Dim decodedArgs = ArrayBuilder(Of TypeSymbol).GetInstance(typeArgs.Length)
+            Dim anyDecoded = False
+            ' Visit the type arguments in reverse
+            For i As Integer = typeArgs.Length - 1 To 0 Step -1
+                Dim typeArg = typeArgs(i)
+                Dim decoded = DecodeType(typeArg)
+                anyDecoded = anyDecoded Or Not ReferenceEquals(decoded, typeArg)
+                decodedArgs.Add(decoded)
+            Next
+
+            If Not anyDecoded Then
+                decodedArgs.Free()
+                Return typeArgs
+            End If
+
+            decodedArgs.ReverseContents()
+            Return decodedArgs.ToImmutableAndFree()
+        End Function
+
+        Private Function DecodeArrayType(type As ArrayTypeSymbol) As ArrayTypeSymbol
+            Dim decodedElementType = DecodeType(type.ElementType)
+            Return If(decodedElementType Is type,
+                        type,
+                        If(type.IsSZArray,
+                            ArrayTypeSymbol.CreateSZArray(decodedElementType,
+                                                          type.CustomModifiers,
+                                                          _containingAssembly),
+                            ArrayTypeSymbol.CreateMDArray(decodedElementType,
+                                                          type.CustomModifiers,
+                                                          type.Rank,
+                                                          type.Sizes,
+                                                          type.LowerBounds,
+                                                          _containingAssembly)))
+
+
+        End Function
+
+        Private Function EatElementNamesIfAvailable(numberOfElements As Integer) As ImmutableArray(Of String)
+            Debug.Assert(numberOfElements > 0)
+
+            ' If we don't have any element names there's nothing to eat
+            If _elementNames.IsDefault Then
+                Return _elementNames
+            End If
+
+            ' We've gone past the end of the names -- bad metadata
+            If numberOfElements > _namesIndex Then
+                Throw New InvalidOperationException()
+            End If
+
+            ' Check to see if all the elements are null
+            Dim start = _namesIndex - numberOfElements
+            Dim allNull = True
+
+            For i As Integer = 0 To numberOfElements - 1
+                If _elementNames(start + i) IsNot Nothing Then
+                    allNull = False
+                End If
+            Next
+
+            If allNull Then
+                _namesIndex = start
```

Comment:
Assignment can be moved to line 285 and redundant assignment at 304 can be removed.